elastic/cloud-on-k8s
# Allow K8S node labels to be propagated as pod annotations

Fixes https://github.com/elastic/cloud-on-k8s/issues/3933

## How to test (tested on GKE)

* Start the operator with the following flag: `--exposed-node-labels="topology.kubernetes.io/.*,failure-domain.beta.kubernetes.io/.*"`
* Annotate an Elasticsearch deployment: `k annotate --overwrite es/elasticsearch-sample   eck.k8s.elastic.co/downward-node-labels="topology.kubernetes.io/zone,topology.kubernetes.io/region"`
* Wait for the Pods to be restarted
* Check the annotations: `k get pod -l common.k8s.elastic.co/type=elasticsearch -o go-template='{{range .items}}{{.metadata.annotations}}{{printf "\n"}}{{end}}'`

To get the logs of the init script you can run `k logs  elasticsearch-sample-es-default-0 -c elastic-internal-init-filesystem  -f`

## Misc

* I eventually decided to reuse the `elastic-internal-init-filesystem` init container to wait for the annotations to be set.
* This PR should not restart the current `RUNNING` Pods. The `elastic-internal-init-filesystem` init-container should be left untouched as long as the feature is not used on an Elasticsearch resource.
* Pods are patched early in the reconciliation loop. I'm not sure if there's an ideal place (or a place we should avoid)
* A Pod is considered as scheduled when both its `Spec.NodeName` has a non-empty value and the relevant condition `PodScheduled` is set to `True` (see K8S binding code [here](https://github.com/kubernetes/kubernetes/blob/6c357f9996070fd015e08eec24f07c82b4b8b4a6/pkg/registry/core/pod/storage/storage.go#L217-L230))
* The primary intent for this PR is to propagate _topology_ node labels, hence node labels are considered as immutable and K8S `Nodes` are not watched.
* When an Elasticsearch resource is submitted the admission controller only check that the node label is allowed. Node labels are not checked until the Pod is scheduled. This is mostly because node labels may be inconsistent.
* I'll try to think about some e2e tests in a subsequent PR.

## To be addressed in follow up PRs

- [ ] Update documentation, including `docs/operating-eck/operating-eck.asciidoc`
- [ ] Enable telemetry
